src/main/java/edu/hm/hafner/analysis/ast/Ast.java

```diff
@@ -0,0 +1,525 @@
+            }
+            if (node.getNextSibling() != null) {
+                collectElementsOfLine(node.getNextSibling(), line, elements);
+            }
+        }
+    }
+
+    /**
+     * Runs entirely through the AST.
+     *
+     * @param node
+     *         Expects the root of the AST which is run through
+     */
+    public void runThroughAST(final DetailAST node) {
+        if (node != null) {
+            allElements.add(node);
+            isConstant(node);
+
+            if (node.getFirstChild() != null) {
+                runThroughAST(node.getFirstChild());
+            }
+            if (node.getNextSibling() != null) {
+                runThroughAST(node.getNextSibling());
+            }
+        }
+    }
+
+    /**
+     * Runs entirely through the AST.
+     *
+     * @param node
+     *         Expects the root of the AST which is run through
+     */
+    public void calcConstants(final DetailAST node) {
+        if (node != null) {
+            isConstant(node);
+
+            if (node.getFirstChild() != null) {
+                calcConstants(node.getFirstChild());
+            }
+            if (node.getNextSibling() != null) {
+                calcConstants(node.getNextSibling());
+            }
+        }
+    }
+
+    /**
+     * Choose the Area around the warning.
+     *
+     * @return the Area
+     */
+    public abstract List<DetailAST> chooseArea();
+
+    /**
+     * Depicts the result of chooseArea() as a string.
+     *
+     * @param delimiter
+     *         the delimiter between the ast-elements. If delimiter is equal z, then no delimiter would be inserted.
+     *
+     * @return the result in string-format
+     */
+    public String chosenAreaAsString(final char delimiter) {
+        StringBuilder stringBuilder = new StringBuilder();
+
+        for (DetailAST detailAST : chooseArea()) {
+            stringBuilder.append(TokenUtil.getTokenName(detailAST.getType()));
+            stringBuilder.append(delimiter);
+        }
+
+        return stringBuilder.toString();
+    }
+
+    /** Necessary for ASTs with name. */
+    private String name = "";
+
+    /**
+     * Sets the name to the specified value.
+     *
+     * @param name
+     *         the value to set
+     */
+    public void setName(final String name) {
+        this.name = name;
+    }
+
+    /**
+     * Returns the name.
+     *
+     * @return the name
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * Returns the hash code of the selected part of the AST. The hash code is derived from the SHA1 digest.
+     *
+     * @return hash code of the AST
+     */
+    public long getContextHashCode() {
+        return SHA_1_TO_LONG_CONVERTER.toLong(getDigest());
+    }
+
+    /**
+     * Returns a digest of the selected part of the AST.
+     *
+     * @return the digest of the AST, represented as SHA1 hash code
+     */
+    public String getDigest() {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'getDigest()' has a cyclomatic complexity of 11.](https://app.codacy.com/manual/uhafner/analysis-model/pullRequest?prid=5019963)